### PR TITLE
Feature/usbl

### PR DESCRIPTION
**Pull request by rsaez-itaca**
Integration of Seatrac in the USBL feature.
Extended IMC messages need attention because  the Seatrac modem can't fill all fields. 
This pull request is intented for code review. Please don't accept it, this development needs more testing.

### CODE REVIEW

**Comment by josebraga** [comment]
Thanks for the effort on this, it's certainly much appreciated :)
I'll take a look on this tomorrow and on next week (the review part)

I still have to do the experimental tests on the "original" feature/usbl (as I haven't even connected with the modems yet!). So, if in the meantime you go ahead with tests please give me feedback on those as you'll be leading this.

Thanks
José

**Comment by rsaez-itaca** [comment]
Dear José,
Of course i will give you feeback. For now, i can say you that the discovery phase works OK.
We will need to test the scheduling phase with two or three target modem, but the target selection part and the timeout part work OK with one target. 
On monday, i'm testing the ranging and the quick reply.

Thank you for your quick reply.
Raúl.

**Comment by rsaez-itaca** [comment]
Dear José,
Today I have more feedback.
The ranging and the quick reply work OK.
Now, I'm testing the change of parameters from Neptus:
- Enable/disable the node: works.
- Set/unset fix request: works.
- Set/unset quick reply: works.
- Change the period: needs more test.

**Comment by josebraga** [comment]
those are great news!!
thank you!

**Comment by rsaez-itaca** [comment]
Dear José,
I think that the whole system is working as expected with the last fix. 
Comment:
The node has a delayed reaction when a parameter is changed from Neptus. This delay is due to the flooding protection. 
Question:
I think errors in remote systems need attention. For example, if a target is out of range or in error state, the usbl modem continues sending messages indefinitely (wasting energy unnecessarily). Regarding nodes, they wait usbl modem messages indefinitely although the modem is dead, but they could search another usbl in range  (e.g in a scenario with usbl cells). What do you think about?

**Comment by josebraga** [comment]
Great bug catches - I can see your tests were very thorough, which isn't easy when using other group's software tools. Congrats and thanks!

I cherry-picked all UsblTools.hpp commits to my branch feature/usbl. 
The rest of your branch will be merged directly to the master once I merge the feature/usbl to the master first (which I am expecting to do before July)

If it's in your interest to do some sea trials using the LAUV systems (either using evologics or seatrac modems), then, by all means, please do so - just please generate a branch named 'mission/_something_'  from your developments, merge master and generate IMC.

I totally agree with your question. Since I coded without doing any tests I didn't want to risk making it too complex before testing. For the first example probably I would have the modem "sleep" (by increasing the timer) and occasionally ping, or restart immediately if any messages from the target are received. In the second example, I guess I could request again after some downtime.. but **from what I remember** (I am not sure..) I have no guard if two usbl modems reply to the request.. they will end competing with each other - since it's not common to have more than USBL modems in the water (at least in our setups) I didn't worry too much with this yet. But I have no idea if the best is to just use both information, and apply some TDMA, or just choose the best (nearest or best rssi) option.

**Comment by josebraga** [comment]
I meant: 
they will end **up** competing with each other - but, since it's not common to have more than **one** USBL modem in the water (at least in our setups) I didn't worry too much with this yet. But I have no idea if the best solution is to use both (applying some TDMA) or just choose one (nearest or best rssi).

**Comment by rsaez-itaca** [comment]
mmm, for the first example, i think that removing the target of the list when it doesn't respond a specific number of times is enough. For the second example, i didn't contemplate two modem in the water (because the code doesn't consider this case from the begining). I thought about a scenario with usbl cells separated a few tens of meters each other and the AUV navigating between cells. Ok, I can give you another example for the second question. What happens if the system of the modem USBL has an error and the target list is erased? Nodes wait indefinitly...
I think that the solution to both problems shouldn't be complicated, only a counter and a threshold to restart.
Two usbl in water is another question and the solution is not trivial. I think that you can face this question when the system works OK in all cases with one usbl.

**Comment by rsaez-itaca** [comment]
Ok, after this last fix, I send you a tested and simple solution (based on counters and thresholds) to resolve the discussed issues.

**Comment by josebraga** [comment]
Hi Raul,

Just to make sure I understood correctly.
The node will restart requesting if it is more than 5 (c_max) times period
without getting fixes from modem.
Each target now has an error count variable and hasFailed that increments
and checks if threshold is above *c_max*

The modem will call target->hasFailed at each overflow, and if true, remove
target. If target replies in the meantime, the
modem erases target's error counter.

I like it, thanks


2016-06-01 11:50 GMT+01:00 Raúl Sáez Cañete <notifications@github.com>:

**Comment by rsaez-itaca** [comment]
Yes, it's as you say. Please, allow me to do a consistency fix.

Oks, I'm following you advice about the branch for trials.
Best regards.

**Comment by rsaez-itaca** [comment]
Dear José,

Last fixes are important to ensure a correct behaviour of the system. Please, check them when you have free time.

Best regards,
Raúl.

**Comment by josebraga** [comment]
Hi Raul,

just got back from holidays
I was merging your modifications but I am wondering about:

> commit bc3fe1bd48f549009c165024eb7f017d65b81c22
> Author: Raúl Sáez <rsaez@itaca.upv.es>
> Date:   Tue Jun 14 18:41:07 2016 +0100
> 
>     Fixed a bug that produces unexpected resets of the error count in nodes when promiscuous mode is enabled.

I am not so sure about it because the way I am seeing it, IF the node is able to listen to messages from the system (even if not addressed to him), this should mean the modem is alive. The risk is that though the modem is alive and communicating with other nodes it may have removed the "node" from its list and thus will never reply to him, while the node will wait forever since it's resetting timer ?

Thanks
José

**Comment by josebraga** [comment]
By the way.
I added UsblConfig and UsblModem in an IMC branch: https://github.com/LSTS/imc/compare/feature/usblmodem

I modified UsblTools to reflect those modifications: https://github.com/LSTS/dune/commit/cef03fc4d10e80dbf67d8b41781df02aeb0a174f

Thanks
José

**Comment by rsaez-itaca** [comment]
> The risk is that though the modem is alive and communicating with other nodes it may have removed the "node" from its list and thus will never reply to him, while the node will wait forever since it's resetting timer ?

Yes, this is one risk. Another risk appears when we consider range extension for USBL. The node navigates until it exits the range of USBL1, but it is in the range of USBL2. The node listens another modem messages and reset its timer, then it will never connect with the USBL2.  On the other hand, it's a consistency problem too. The node waits for USBL messages, it doesn't mind other modem messages to perform the usbl task (only to collect more information). Then, it mustn't change the usbl task state (or reset the error count) when listen other node messages although there is no risk.

**Comment by josebraga** [comment]
Understood. I had already merged everything to feature/usbl :)

By the way, the "seatrac" part is finished and ready for merge?

thanks

**Comment by josebraga** [comment]
Thanks Rául!

@JoaoTeixiera: can you please review seatrac part? 

once this is reviewed I'll add to mainline.

**Comment by rsaez-itaca** [comment]
Thank you for allowing me to contribute in your development.

**Comment by josebraga** [comment]
Well, I guess it's more the other way around, thank you for helping us !

File: src/Transports/Seatrac/Task.cpp [hunk 1]
```diff
@@ -627,11 +622,6 @@ namespace Transports
               {
                 sendProtectedCommand(commandCreateSeatrac(CID_DAT_SEND, m_data_beacon));
               }
-              else
-              {
-                // Data communication done
-                clearTicket(IMC::UamTxStatus::UTS_DONE);
```
Comment (rsaez-itaca) [inline src/Transports/Seatrac/Task.cpp]: @JoaoTeixiera For future review: I have tested this line while I was testing the usbl task (Both manta-12 and caravela) and I didn't get any core dump into log folder.

**Comment by rsaez-itaca** [comment]
@JoaoTeixiera I have had data lost in the reception of CID_STATUS messages when the modem receives an acoustic message. The data lost is not produce into the Seatrac code, it's produced when the handle is read. It's possible that the problem was produced by the serial-tcp converter that Caravela was using, but more tests are necessary to ensure it. You will need to face this in the future then I attach a log for your convenience. The log shows all messages read from the handle i.e. from bfr array: 
[log.txt](https://github.com/LSTS/dune/files/354393/log.txt)

**Pull request closed without merging**